app/models/assignment.rb

```diff
@@ -850,7 +853,67 @@ def get_num_valid
              .count(&:is_valid?)
   end
 
-  def get_num_marked(ta_id = nil)
+  def marked_result_ids_for(ta_id)
+    cache_ta_results
+    @cache_ta_results.dig(ta_id, :marked_result_ids) || []
+  end
+
+  def cache_ta_results
+    return @cache_ta_results if defined? @cache_ta_results
+
+    data = current_results.joins(grouping: :tas).pluck('tas.id', 'results.id', 'results.marking_state')
+    # Group results by TA ID
+    grouped_data = data.group_by { |ta_id, _result_id, _marking_state| ta_id }
+
+    results_by_id = current_results.includes(:marks).index_by(&:id)
+    @cache_ta_results = grouped_data.each_with_object({}) do |(ta_id, results), hash|
+      total_results = results.map { |_, result_id, _| result_id }
+
+      if self.assign_graders_to_criteria
+        # Get the list of criteria IDs assigned to this TA
+        assigned_criteria_ids = self.criteria.flat_map do |criterion|
```

Comment:
This part still has an N+1 query because it's repeated once per TA.

It is safe to pull this out of the loop, and create a `Hash` mapping `ta_id` to the list of criterion ids.

Also, note that you can directly access `self.criterion_ta_associations` (because that's an association defined for the `Assignment` model).